src/SFA.DAS.EAS.Portal/Application/Services/AccountsServiceWithDuplicateCheck.cs

```diff
@@ -0,0 +1,33 @@
+﻿using SFA.DAS.EAS.Portal.Database.Models;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.EAS.Portal.Application.Services
+{
+    public class AccountsServiceWithDuplicateCheck : IAccountsService
+    {
+        private readonly IAccountsService _accountsService;
+        private readonly IMessageContext _messageContext;
+        public AccountsServiceWithDuplicateCheck(IAccountsService accountsService, IMessageContext messageContext)
```

Comment:
This service introduces a new pattern to handle de-dupe of messages (the existing pattern is contained in the `Account` model), so now we have two separate ways of handling de-dupe, one used when handling `ReservationCreatedEvent`'s and one when handling `CohortApprovalRequestedByProvider` events. For this pattern to work, `Account`'s private `JsonConstructor` decorated constructor has had to be made public, which breaks the existing pattern, which relies on constructors setting up the required invariants on the `Account` object.

I'm not opposed to extracting de-duplication out of the `Account` model, but we need consistency in how we handle it across message handlers, and I have concerns about the thread safety of this particular pattern (which I have documented in a separate comment).

Reply:
It's not a new pattern it is decoupling the concerns as per SOLID code principles.  454 will refactor the existing code to match.

Reply:
Out of date now, as both handlers use the new pattern (or close enough!).